crates/iroha_core/src/tx.rs

```diff
@@ -296,3 +303,539 @@ impl StateBlock<'_> {
+        async fn fires_for_each_transaction() {
+            let mut sandbox =
+                Sandbox::new().with_data_trigger_oneshot("bob", "carol", INIT_BALANCE["bob"]);
+            sandbox.transfer_one("alice", "bob");
+            sandbox.transfers_batched(1, "carol", "dave", INIT_BALANCE["bob"]);
+            let mut block = sandbox.block();
+            let _events = block.apply();
+            block.assert_balances([
+                ("alice", -1),
+                ("bob", 1 - INIT_BALANCE["bob"] as i32),
+                ("carol", 0),
+                ("dave", INIT_BALANCE["bob"] as i32),
+            ]);
+        }
+
+        /// # Scenario
+        ///
+        /// 1. Transaction transfers an asset from Alice to Bob twice.
+        /// 2. Trigger __should fire once__ and transfer one from Bob to Carol.
+        #[tokio::test]
+        async fn fires_at_most_once_per_transaction() {
+            let mut sandbox = Sandbox::new().with_data_trigger("bob", "carol");
+            sandbox.transfers_batched(2, "alice", "bob", 1);
+            let mut block = sandbox.block();
+            let _events = block.apply();
+            block.assert_balances([("alice", -2), ("bob", 1), ("carol", 1)]);
+        }
+
+        /// # Scenario
+        ///
+        /// 1. Transaction transfers an asset from Alice to Bob.
+        /// 2. Depth-first trigger chain: Bob -> Carol -> Dave -> Bob.
+        /// 3. Sibling trigger (Bob -> Eve) executes only after the depth-first chain completes.
+        #[tokio::test]
+        #[expect(clippy::cast_possible_wrap)]
+        async fn chains_in_depth_first_order() {
+            let mut sandbox = Sandbox::new()
+                .with_data_trigger_oneshot("bob", "carol", INIT_BALANCE["bob"])
+                // Sibling trigger waits for depth-first resolution.
+                .with_data_trigger_oneshot("bob", "eve", INIT_BALANCE["bob"])
+                .with_data_trigger_oneshot("carol", "dave", INIT_BALANCE["bob"])
+                .with_data_trigger_oneshot("dave", "bob", INIT_BALANCE["bob"]);
+            sandbox.transfer_one("alice", "bob");
+            let mut block = sandbox.block();
+            let _events = block.apply();
+            block.assert_balances([
+                ("alice", -1),
+                ("bob", 1 - INIT_BALANCE["bob"] as i32),
+                ("carol", 0),
+                ("dave", 0),
+                ("eve", INIT_BALANCE["bob"] as i32),
+            ]);
+        }
+
+        /// All or none of the initial transaction and subsequent data triggers should take effect.
+        #[tokio::test]
+        async fn atomically_chains_from_transaction() {
+            aborts_on_execution_error(&TriggerOrigin::ExternalTransaction);
+            aborts_on_exceeding_depth(&TriggerOrigin::ExternalTransaction);
+            commits_on_depleting_lives(&TriggerOrigin::ExternalTransaction);
+            commits_on_regular_success(&TriggerOrigin::ExternalTransaction);
```

Comment:
@s8sato 

Add assertions to verify which events are published, depending on whether the transaction commits or aborts.